network/p2p/gossip/metrics.go

```diff
@@ -0,0 +1,185 @@
+
+	labelsLock        sync.Mutex
+	gossipablesLabels map[ids.ID]string
+}
+
+// NewMetrics returns a common set of metrics
+func NewMetrics(
+	metrics prometheus.Registerer,
+	namespace string,
+) (*Metrics, error) {
+	m := &Metrics{
+		count: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Name:      "gossip_count",
+				Help:      "amount of gossip (n)",
+			},
+			ioTypeDroppedLabels,
+		),
+		bytes: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Name:      "gossip_bytes",
+				Help:      "amount of gossip (bytes)",
+			},
+			ioTypeDroppedLabels,
+		),
+		tracking: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Name:      "gossip_tracking",
+				Help:      "number of gossipables being tracked",
+			},
+			typeLabels,
+		),
+		trackingLifetimeAverage: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "gossip_tracking_lifetime_average",
+			Help:      "average duration a gossipable has been tracked (ns)",
+		}),
+		topValidators: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Name:      "top_validators",
+				Help:      "number of validators gossipables are sent to due to stake",
+			},
+			typeLabels,
+		),
+		gossipablesLabels: make(map[ids.ID]string),
+	}
+	err := errors.Join(
+		metrics.Register(m.count),
+		metrics.Register(m.bytes),
+		metrics.Register(m.tracking),
+		metrics.Register(m.trackingLifetimeAverage),
+		metrics.Register(m.topValidators),
+	)
+	return m, err
+}
+
+func (m *Metrics) observeMessage(labels prometheus.Labels, count int, bytes int) error {
+	countMetric, err := m.count.GetMetricWith(labels)
+	if err != nil {
+		return fmt.Errorf("failed to get count metric: %w", err)
+	}
+
+	bytesMetric, err := m.bytes.GetMetricWith(labels)
+	if err != nil {
+		return fmt.Errorf("failed to get bytes metric: %w", err)
+	}
+
+	countMetric.Add(float64(count))
+	bytesMetric.Add(float64(bytes))
+	return nil
+}
+
+// ObserveIncomingGossipable allows external-package mempools to label a custom gossipable drop-reason and have it
+// included in the gossip metrics.
+func (m *Metrics) ObserveIncomingGossipable(txID ids.ID, label string) {
+	m.labelsLock.Lock()
+	defer m.labelsLock.Unlock()
+	_, ok := m.gossipablesLabels[txID]
+	if ok {
+		// we already have this gossipable labeled; don't re-label.
+		return
+	}
+	m.gossipablesLabels[txID] = label
+}
+
+// observeReceivedMessage summarizes the metrics for a given received gossip message and update the metrics accordingly.
+func (m *Metrics) observeReceivedMessage(typeLabel string, gossipables map[ids.ID]int, malformedBytes int, malformedCount int) error {
+	labelBytes := make(map[string]int, len(gossipables))
+	labelCount := make(map[string]int, len(gossipables))
+
+	labelBytes[droppedMalformed] = malformedBytes
+	labelCount[droppedMalformed] = malformedCount
+
+	m.labelsLock.Lock()
```

Comment:
Why don't we just lock at the start of this function and use `defer` to unlock? Unless we have data to indicate that lock contention is an issue I don't think it's worth making the code more confusing to read.

Reply:
ok, done.